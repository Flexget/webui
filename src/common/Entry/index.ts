import EntryCard from './cards';

export default EntryCard;
