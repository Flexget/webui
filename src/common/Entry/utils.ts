import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
