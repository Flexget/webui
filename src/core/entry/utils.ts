import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
