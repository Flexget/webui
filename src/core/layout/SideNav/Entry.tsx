import React, { ComponentType, FC } from 'react';
import { css } from '@emotion/core';
import { Link } from 'react-router-dom';
import { ListItem, ListItemText, ListItemIcon, Theme } from '@material-ui/core';

interface Props {
  onClick?: () => void;
  path?: string;
  Icon: ComponentType;
  className?: string;
  name: string;
}

export const colorClass = (theme: Theme) => css`
  color: ${theme.palette.secondary.light};
`;

const navItem = (theme: Theme) => css`
  ${colorClass(theme)};
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid ${theme.palette.primary.main};
  }
`;

const SideNavEntry: FC<Props> = ({ onClick, path, Icon, name, className }) => {
  const item = (
    <ListItem css={theme => [navItem(theme), className]} onClick={onClick}>
      <ListItemIcon css={colorClass}>
        <Icon />
      </ListItemIcon>
      <ListItemText css={colorClass} disableTypography primary={name} />
    </ListItem>
  );

  if (path) {
    return <Link to={path}>{item}</Link>;
  }

  return item;
};

export default SideNavEntry;
