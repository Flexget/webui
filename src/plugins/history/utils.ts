import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
