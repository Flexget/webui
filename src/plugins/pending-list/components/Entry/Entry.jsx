import React from 'react';
import PropTypes from 'prop-types';
import TextTruncate from 'react-text-truncate';
import { Textfit } from 'react-textfit';
import Button from 'material-ui/Button';
import { CardHeader, CardActions, CardMedia, CardContent } from 'material-ui/Card';
import Tooltip from 'material-ui/Tooltip';
import Chip from 'material-ui/Chip';
import IconButton from 'material-ui/IconButton';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { EntryShape } from 'plugins/pending-list/data/shapes';
import { removeDupesIgnoreCase } from 'utils';
import { getValues } from 'utils/object';
import { ActionIcon, content, subheader, EntryCard } from './styles';


/* eslint-disable camelcase */
class Entry extends React.PureComponent {
  static propTypes = {
    entry: EntryShape.isRequired,
    openRemoveModal: PropTypes.func.isRequired,
    approveEntry: PropTypes.func.isRequired,
    rejectEntry: PropTypes.func.isRequired,
  };

  static fieldMap = {
    genres: ['imdb_genres', 'tvdb_genres', 'tvmaze_genres', 'trakt_genres', 'tmdb_genres'],
    poster: ['imdb_photo', 'tvdb_posters', 'tvmaze_series_original_image', 'tmdb_posters'],
    description: ['imdb_plot_outline', 'tvdb_overview', 'tmdb_tagline', 'trakt_series_overview', 'tvmaze_series_summary'],
  }

  cleanTitle() {
    const { entry: { title, entry } } = this.props;

    if (entry.series_name) {
      const epId = entry.series_id ? ` ${entry.series_id}` : '';
      return `${entry.series_name}${epId}`;
    }

    if (entry.movie_name) {
      const year = entry.movie_year ? ` (${entry.movie_year})` : '';
      return `${entry.movie_name}${year}`;
    }
    return title;
  }

  qualityChips() {
    const chips = [];
    const { entry: { id, entry: { quality } } } = this.props;
    if (quality) {
      Object.entries(quality.split(' ')).forEach(([idx, q]) => {
        const chipId = `${id}-${idx}-quality`;
        chips.push(<Chip label={q} id={chipId} style={{ marginRight: 4, marginBottom: 2 }} />);
      });
    }
    return chips;
  }

  genreChips() {
    const { entry: { id, entry } } = this.props;

    const genres = getValues(entry, Entry.fieldMap.genres);
    const uniqueGenres = removeDupesIgnoreCase([].concat(...genres));

    const chips = [];
    Object.entries(uniqueGenres).forEach(([idx, genre]) => {
      const chipId = `${id}-${idx}-genre`;
      chips.push(<Chip label={genre} id={chipId} style={{ marginRight: 4, marginBottom: 2 }} />);
    });
    return chips;
  }

  subHeader() {
    const qualityChips = this.qualityChips();
    const genreChips = this.genreChips();

    return (
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {qualityChips}
        {genreChips}
      </div>
    );
  }

  posters() {
    const { entry: { entry } } = this.props;
    const posters = [];
    Entry.fieldMap.poster.forEach((field) => {
      let values = entry[field];
      if (values) {
        if (!Array.isArray(values)) values = [values];
        posters.push(...values);
      }
    });
    return posters;
  }

  descriptions() {
    const { entry: { entry } } = this.props;
    const descriptions = [];
    Entry.fieldMap.description.forEach((field) => {
      let values = entry[field];
      if (values) {
        if (!Array.isArray(values)) values = [values];
        descriptions.push(...values);
      }
    });
    return descriptions;
  }

  links() {
    const { entry: { entry } } = this.props;
    const links = [];
    if ('imdb_url' in entry) {
      links.push(<Button variant="fab" color="primary" aria-label="add" href={entry.imdb_url}>IMDB</Button>);
    }
    if ('tvdb_url' in entry) {
      links.push(<Button variant="fab" color="primary" aria-label="add" href={entry.tvdb_url}>TVDB</Button>);
    }
    return links;
  }

  render() {
    const { entry: { title, approved } } = this.props;
    const { openRemoveModal, approveEntry, rejectEntry } = this.props;

    const titleClean = this.cleanTitle();
    const posters = this.posters();
    const descriptions = this.descriptions();
    const links = this.links();

    return (
      <EntryCard>
        {posters.length > 0 &&
        <CardMedia style={{ width: 250, height: 220 }} image={posters[0]} title={titleClean} />
        }
        <div style={{ display: 'flex', flexDirection: 'column' }}>
          <CardHeader
            title={(
              <Textfit max={25} min={20} mode="multi">
                <Tooltip title={title}><span>{titleClean}</span></Tooltip>
              </Textfit>
            )}
            subheader={this.subHeader()}
            action={
              <IconButton onClick={openRemoveModal}>
                <FontAwesomeIcon icon={['far', 'trash-alt']} />
              </IconButton>
            }
            classes={{
              content,
              subheader,
            }}
          />
          {
            descriptions.length > 0 &&
            <CardContent>
              <TextTruncate
                line={3}
                truncateText="â€¦"
                text={descriptions[0]}
              />
            </CardContent>
          }
          {
            links.length > 0 &&
            <CardContent>
              {links}
            </CardContent>
          }
          <CardActions>
            {
              approved ? (
                <Button onClick={rejectEntry}>
                  <ActionIcon icon="times" />
                  Reject
                </Button>
              ) : (
                <Button onClick={approveEntry}>
                  <ActionIcon icon="check" />
                  Approve
                </Button>
              )
            }
          </CardActions>
        </div>
      </EntryCard>
    );
  }
}

export default Entry;
