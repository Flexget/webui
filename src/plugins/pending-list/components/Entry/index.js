import { connect } from 'react-redux';
import { APPROVE_ENTRY, REJECT_ENTRY, REMOVE_ENTRY } from 'plugins/pending-list/data/actions';
import { request } from 'utils/actions';
import Entry from './Entry';

function mapDispatchToProps(dispatch, { entry }) {
  return {
    approveEntry: () => dispatch(request(APPROVE_ENTRY, { entry })),
    removeEntry: () => dispatch(request(REMOVE_ENTRY, { entry })),
    rejectEntry: () => dispatch(request(REJECT_ENTRY, { entry })),
  };
}

export default connect(null, mapDispatchToProps)(Entry);
