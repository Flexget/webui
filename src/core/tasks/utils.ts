import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
