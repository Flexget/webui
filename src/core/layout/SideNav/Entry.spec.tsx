import React from 'react';
import renderer from 'react-test-renderer';
import { themed, router } from 'utils/tests';
import SVGIcon from '@material-ui/core/SvgIcon';
import Entry from './Entry';

describe('core/layout/SideNavEntry', () => {
  it('renders correctly with link', () => {
    const tree = renderer
      .create(router(themed(<Entry Icon={SVGIcon} name="Test" path="/test" />)))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without link', () => {
    const tree = renderer.create(themed(<Entry Icon={SVGIcon} name="Test" />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
