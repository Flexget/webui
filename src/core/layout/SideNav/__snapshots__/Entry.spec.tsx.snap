// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/layout/SideNavEntry renders correctly with link 1`] = `
<a
  href="/test"
  onClick={[Function]}
>
  <li
    className="MuiListItem-root MuiListItem-gutters"
    css={
      Array [
        Object {
          "map": undefined,
          "name": "1fef3mw",
          "next": undefined,
          "styles": "
  
  color: #b0bec5;
;;
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid #ff9800;
  }
",
          "toString": [Function],
        },
        undefined,
      ]
    }
    disabled={false}
  >
    <div
      className="MuiListItemIcon-root"
      css={
        Object {
          "map": undefined,
          "name": "2tnp70",
          "next": undefined,
          "styles": "
  color: #b0bec5;
",
          "toString": [Function],
        }
      }
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      />
    </div>
    <div
      className="MuiListItemText-root"
      css={
        Object {
          "map": undefined,
          "name": "2tnp70",
          "next": undefined,
          "styles": "
  color: #b0bec5;
",
          "toString": [Function],
        }
      }
    >
      Test
    </div>
  </li>
</a>
`;

exports[`core/layout/SideNavEntry renders correctly without link 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters"
  css={
    Array [
      Object {
        "map": undefined,
        "name": "1fef3mw",
        "next": undefined,
        "styles": "
  
  color: #b0bec5;
;;
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid #ff9800;
  }
",
        "toString": [Function],
      },
      undefined,
    ]
  }
  disabled={false}
>
  <div
    className="MuiListItemIcon-root"
    css={
      Object {
        "map": undefined,
        "name": "2tnp70",
        "next": undefined,
        "styles": "
  color: #b0bec5;
",
        "toString": [Function],
      }
    }
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    />
  </div>
  <div
    className="MuiListItemText-root"
    css={
      Object {
        "map": undefined,
        "name": "2tnp70",
        "next": undefined,
        "styles": "
  color: #b0bec5;
",
        "toString": [Function],
      }
    }
  >
    Test
  </div>
</li>
`;
